### PR TITLE
switch to cdflib to_datetime for timezone fix

### PR DESCRIPTION
**Pull request by JonathonMSmith**
By using the cdflib funtion to convert epoch to datetime issue #35 should be solved.

This is an enhancement. Although performance may be impacted.

Testing:
OS: Debian 10 on WSL, Windows 10
python 3.7.9
Conda Environment

### CODE REVIEW

**Comment by JonathonMSmith** [comment]
Long note here. I removed all the code utilizing the "center_measurement" variable.
I was going to leave this in as it can provide some corrections to the timestamp if there is information in the CDF about the epoch variance. From that information a delta_time was calculated which was a float, so when added to the epoch values they too became floats. But the cdflib function to_datetime did not handle the cnofs_vefi epoch values well if they were floats. I'm not familiar with any datasets that utilize the DELTA_PLUS_VAR and DELTA_MINUS_VAR variable attributes for epoch, but I'm sure it exists. Also nothing currently utilizes the "center_measurement" CDF instance variable. Is this a functionality that should definitely stay in the cdaweb functions?

**Comment by jklenzing** [comment]
Adding some notes here from #38, since I've been using this branch to troubleshoot.
- For C/NOFS IVM, several dates from 1970 appear in the data if `clean_level='none'`.  This appears for multiple files
- This appears to change somewhere between the `set_epoch` function and the output of `to_pysat`
  - Added print statements to check the minimum epoch value and number of points after `cdflib.cdfepoch.to_datetime(xdata)` in `set_epoch` and just before data is returned in `to_pysat`
  - The number of points is unchanged, but the first 18 points (for 1 Jan 2009) have dates from 1970, changing by milliseconds:
```
1970-01-01 00:00:00.000000000      NaN             NaN           NaN  ...                    NaN           NaN            NaN
1970-01-01 00:00:00.000000001      NaN             NaN           NaN  ...                    NaN           NaN            NaN
1970-01-01 00:00:00.000000002      NaN             NaN           NaN  ...                    NaN           NaN            NaN
1970-01-01 00:00:00.000000003      NaN             NaN           NaN  ...                    NaN           NaN            NaN
1970-01-01 00:00:00.000000004
```
- The first realistic time value matches the initial time point in the file (in this case, `2009-01-01 00:00:00.812000000`)
- Note that because the values are NaNs, these 18 points get dropped if the cleaning routine is run.
- Loaded PLP data as a comparison case, but the dates look fine there.  This may be something unique to the IVM files.

File: pysatNASA/instruments/methods/cdaweb.py [hunk 1]
```diff
@@ -222,13 +222,11 @@ def set_epoch(self, x_axis_var):
                         delta_time = (delta_plus_var
                                       - delta_minus_var) / 2.0
```
Comment (aburrell) [inline pysatNASA/instruments/methods/cdaweb.py:223]: This variable is no longer used anywhere.

Comment (JonathonMSmith) [reply pysatNASA/instruments/methods/cdaweb.py:223]: That's true. After the last discussion I heard about this, It was my understanding that I was going to attempt to convert this to a datetime to be added to the result of cdflib.cdfepoch.to_datetime(xdata).  Not sure if that's correct or not. The hold up is that I haven't figured out if this variable always has the same initial units or not so I haven't pushed through a fix. I had previously removed this variable and all of the code that goes into determining delta_plus_var and delta_minus_var, but that removed a lot of code

Comment (aburrell) [reply pysatNASA/instruments/methods/cdaweb.py:223]: I am trying to do a CINDI analysis at the moment and CDFLIB appears to be broken for our implementation.  May or may not be related to this, but I am trying to fix things.

Comment (aburrell) [reply pysatNASA/instruments/methods/cdaweb.py:223]: What are the units of this variable?

Comment (aburrell) [reply pysatNASA/instruments/methods/cdaweb.py]: Nevermind, figured it out.

**Comment by aburrell** [comment]
Ok, tests are passing now, but I am getting JUNK locally.  I checked out cdflib to try and fix things, since the errors were being raised there.  You can see my branch here: https://github.com/aburrell/cdflib/tree/epoch_array_bugfix

Running this example
```
import datetime as dt
import pysat

ctime = dt.datetime(2014, 1, 4)
cindi = pysat.Instrument('cnofs', 'ivm', clean_level='clean', update_files=True)
cindi.load(date=ctime)
```
yields this error with my new branch because it looks like cdflib's `varget` gets a lot of garbage:

```
---------------------------------------------------------------------------
InvalidIndexError                         Traceback (most recent call last)
<ipython-input-76-750decca70af> in <module>
----> 1 cindi.load(date=ctime)

~/Programs/Git/pysat/pysat/_instrument.py in load(self, yr, doy, end_yr, end_doy, date, end_date, fname, stop_fname, verifyPad)
   2789         else:
   2790             self.data, meta = self._load_data(date=self.date, fid=self._fid,
-> 2791                                               inc=self.load_step)
   2792             if not self.empty:
   2793                 self.meta = meta

~/Programs/Git/pysat/pysat/_instrument.py in _load_data(self, date, fid, inc)
   1224                 data, mdata = self._load_rtn(load_fname, tag=self.tag,
   1225                                              inst_id=self.inst_id,
-> 1226                                              **load_kwargs)
   1227 
   1228                 # ensure units and name are named consistently in new Meta

~/Programs/Git/pysatNASA/pysatNASA/instruments/methods/cdaweb.py in load(fnames, tag, inst_id, file_cadence, flatten_twod)
    496 
    497                     # Select data from multi-day down to daily
--> 498                     data = data.loc[date:date + dt.timedelta(days=1)
    499                                     - dt.timedelta(microseconds=1), :]
    500                     ldata.append(data)

~/Programs/Git/pysatNASA/pysatNASA/instruments/methods/cdaweb.py in to_pysat(self, flatten_twod, labels)
    426                     data[varname] = df
    427 
--> 428         data = pds.DataFrame(data)
    429 
    430         return data, meta

~/Library/Python/3.7/lib/python/site-packages/pandas-0.24.2-py3.7-macosx-10.14-x86_64.egg/pandas/core/frame.py in __init__(self, data, index, columns, dtype, copy)
    390                                  dtype=dtype, copy=copy)
    391         elif isinstance(data, dict):
--> 392             mgr = init_dict(data, index, columns, dtype=dtype)
    393         elif isinstance(data, ma.MaskedArray):
    394             import numpy.ma.mrecords as mrecords

~/Library/Python/3.7/lib/python/site-packages/pandas-0.24.2-py3.7-macosx-10.14-x86_64.egg/pandas/core/internals/construction.py in init_dict(data, index, columns, dtype)
    210         arrays = [data[k] for k in keys]
    211 
--> 212     return arrays_to_mgr(arrays, data_names, index, columns, dtype=dtype)
    213 
    214 

~/Library/Python/3.7/lib/python/site-packages/pandas-0.24.2-py3.7-macosx-10.14-x86_64.egg/pandas/core/internals/construction.py in arrays_to_mgr(arrays, arr_names, index, columns, dtype)
     49     # figure out the index, if necessary
     50     if index is None:
---> 51         index = extract_index(arrays)
     52     else:
     53         index = ensure_index(index)

~/Library/Python/3.7/lib/python/site-packages/pandas-0.24.2-py3.7-macosx-10.14-x86_64.egg/pandas/core/internals/construction.py in extract_index(data)
    310 
    311         if have_series or have_dicts:
--> 312             index = _union_indexes(indexes)
    313 
    314         if have_raw_arrays:

~/Library/Python/3.7/lib/python/site-packages/pandas-0.24.2-py3.7-macosx-10.14-x86_64.egg/pandas/core/indexes/api.py in _union_indexes(indexes, sort)
    181         else:
    182             for other in indexes[1:]:
--> 183                 result = result.union(other)
    184             return result
    185     elif kind == 'array':

~/Library/Python/3.7/lib/python/site-packages/pandas-0.24.2-py3.7-macosx-10.14-x86_64.egg/pandas/core/indexes/base.py in union(self, other, sort)
   2307             this = self.astype('O')
   2308             other = other.astype('O')
-> 2309             return this.union(other, sort=sort)
   2310 
   2311         # TODO(EA): setops-refactor, clean all this up

~/Library/Python/3.7/lib/python/site-packages/pandas-0.24.2-py3.7-macosx-10.14-x86_64.egg/pandas/core/indexes/base.py in union(self, other, sort)
   2330                 result.extend([x for x in rvals if x not in value_set])
   2331         else:
-> 2332             indexer = self.get_indexer(other)
   2333             indexer, = (indexer == -1).nonzero()
   2334 

~/Library/Python/3.7/lib/python/site-packages/pandas-0.24.2-py3.7-macosx-10.14-x86_64.egg/pandas/core/indexes/base.py in get_indexer(self, target, method, limit, tolerance)
   2738 
   2739         if not self.is_unique:
-> 2740             raise InvalidIndexError('Reindexing only valid with uniquely'
   2741                                     ' valued Index objects')
   2742 

InvalidIndexError: Reindexing only valid with uniquely valued Index objects
```

Meanwhile, the retrieved data looks identical using the master branch for cdflib, and without my bugfix it yields:
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-82-750decca70af> in <module>
----> 1 cindi.load(date=ctime)

~/Programs/Git/pysat/pysat/_instrument.py in load(self, yr, doy, end_yr, end_doy, date, end_date, fname, stop_fname, verifyPad)
   2789         else:
   2790             self.data, meta = self._load_data(date=self.date, fid=self._fid,
-> 2791                                               inc=self.load_step)
   2792             if not self.empty:
   2793                 self.meta = meta

~/Programs/Git/pysat/pysat/_instrument.py in _load_data(self, date, fid, inc)
   1224                 data, mdata = self._load_rtn(load_fname, tag=self.tag,
   1225                                              inst_id=self.inst_id,
-> 1226                                              **load_kwargs)
   1227 
   1228                 # ensure units and name are named consistently in new Meta

~/Programs/Git/pysatNASA/pysatNASA/instruments/methods/cdaweb.py in load(fnames, tag, inst_id, file_cadence, flatten_twod)
    495                     data, meta = cdf.to_pysat(flatten_twod=flatten_twod)
    496 
--> 497                     # Select data from multi-day down to daily
    498                     data = data.loc[date:date + dt.timedelta(days=1)
    499                                     - dt.timedelta(microseconds=1), :]

~/Programs/Git/pysatNASA/pysatNASA/instruments/methods/cdaweb.py in __init__(self, filename, varformat, var_types, center_measurement, raise_errors, regnames, datetime, **kwargs)
    104                                 + self._cdf_info['zVariables'])
    105 
--> 106         self.load_variables()
    107 
    108     def __enter__(self):

~/Programs/Git/pysatNASA/pysatNASA/instruments/methods/cdaweb.py in load_variables(self)
    333                         ydata[ydata == var_atts["FILLVAL"]] = np.nan
    334 
--> 335             index = self.get_index(variable_name)
    336 
    337             try:

~/Programs/Git/pysatNASA/pysatNASA/instruments/methods/cdaweb.py in get_index(self, variable_name)
    260         elif "DEPEND_0" in var_atts:
    261             x_axis_var = var_atts["DEPEND_0"]
--> 262             self.set_epoch(x_axis_var)
    263 
    264         dependencies = []

~/Programs/Git/pysatNASA/pysatNASA/instruments/methods/cdaweb.py in set_epoch(self, x_axis_var)
    220                 if self._datetime:
    221                     # Convert xdata to numpy datetime
--> 222                     new_xdata = cdflib.cdfepoch.to_datetime(xdata)
    223 
    224                     # Add delta to time, if both plus and minus are defined

~/Programs/Git/cdflib/cdflib/epochs.py in to_datetime(cls, cdf_time, to_np)
    187         If to_np is True, then the values will be returned in a numpy array.
    188         """
--> 189         times = cls.breakdown(cdf_time, to_np=True)
    190         times = np.atleast_2d(times)
    191         times = cls._compose_date(*times.T)

~/Programs/Git/cdflib/cdflib/epochs.py in breakdown(epochs, to_np)
    152             return CDFepoch.breakdown_tt2000(epochs, to_np)
    153         elif isinstance(item, (float, np.float64)):
--> 154             return CDFepoch.breakdown_epoch(epochs, to_np)
    155         elif isinstance(item, (complex, np.complex128)):
    156             return CDFepoch.breakdown_epoch16(epochs, to_np)

~/Programs/Git/cdflib/cdflib/epochs.py in breakdown_epoch(epochs, to_np)
   1478                 hour_AD = minute_AD / 60.0
   1479                 day_AD = hour_AD / 24.0
-> 1480                 jd = int(1721060 + day_AD)
   1481                 l = jd+68569
   1482                 n = int(4*l/146097)

TypeError: only size-1 arrays can be converted to Python scalars
```

**Comment by jklenzing** [comment]
@aburrell, the cindi files for Jan 4-31 2014 have broken dates.  This is a problem with the files on spdf (as well as the originals).   

@JonathonMSmith, I remember you ran into this issue with your previous analysis.  Do you remember if pysatCDF would error on these files or return an empty dataset?

**Comment by jklenzing** [comment]
Found it: pysat/pysat#87

We wound up solving this by dropping bad datetimes.

File: pysatNASA/instruments/methods/cdaweb.py [hunk 2]
```diff
@@ -212,23 +215,27 @@ def set_epoch(self, x_axis_var):
...
-
-        if self.get_dependency(x_axis_var) is None:
             if ('CDF_TIME' in data_type_description) or \
                     ('CDF_EPOCH' in data_type_description):
-                xdata = cdflib.cdfepoch.unixtime(xdata)
-                xdata = np.array(xdata) + delta_time
                 if self._datetime:
-                    xdata = pds.to_datetime(xdata, unit='s')
+                    # Convert xdata to datetime
+                    new_xdata = cdflib.cdfepoch.to_datetime(xdata)
```
Comment (jklenzing) [inline pysatNASA/instruments/methods/cdaweb.py:222]: This line should probably be surrounded by a `try` / `except(TypeError)` statement to catch the error in the Jan 4-31 2014 cindi data.  The question is what do we want the output to be?  Raise a more useful error message? Warn and drop the data?

Comment (aburrell) [reply pysatNASA/instruments/methods/cdaweb.py:222]: Warn and drop: something about "File has corrupted times, contact CDAWeb for them to fix it"

Comment (aburrell) [reply pysatNASA/instruments/methods/cdaweb.py:222]: Probably also worth adding a try/except to line 428, in case my branch for cdflib gets merged in.

**Comment by aburrell** [comment]
@jklenzing if the cindi files for Jan 4-31 2014 have broken dates and aren't going to be fixed, perhaps we should find a way to blacklist them in the CINDI download and load routines.  That would save computing time.

File: pysatNASA/instruments/methods/cdaweb.py [hunk 3]
```diff
@@ -482,24 +494,28 @@ def load(fnames, tag=None, inst_id=None, file_cadence=dt.timedelta(days=1),
...
                     ldata.append(temp_data)
 
         # Combine individual files together
-        data = pds.concat(ldata)
+        if len(ldata) > 0:
+            try:
+                data = pds.concat(ldata)
+            except InvalidIndexError as ierr:
+                wmsg = "Invalid times in data file(s): {:}".format(str(ierr))
+                warnings.warn(wmsg)
```
Comment (aburrell) [inline pysatNASA/instruments/methods/cdaweb.py:517]: This should probably be a logger warning.

**Comment by jklenzing** [comment]
> @jklenzing if the cindi files for Jan 4-31 2014 have broken dates and aren't going to be fixed, perhaps we should find a way to blacklist them in the CINDI download and load routines. That would save computing time.

I've thougth about this as well, but I have mixed feelings about it.  In theory, the files should be fixed, and I'd prefer to have an adaptive routine that can recognize bad files and let people know should this happen again with another data set.

**Comment by aburrell** [comment]
> I've thought about this as well, but I have mixed feelings about it. In theory, the files should be fixed, and I'd prefer to have an adaptive routine that can recognize bad files and let people know should this happen again with another data set.

I created an issue over at pysat for future discussion: https://github.com/pysat/pysat/issues/713

I agree that it is important for pysatNASA to handle the errors that have been raised here, but it does slow the processing down and since NASA doesn't seem to care that there are bad files on the server, this could improve people's lives.

**Comment by aburrell** [comment]
The 1970 issue was created by assuming that all the CDF pandas Series used a DatetimeIndex when they actually used a mix of DatetimeIndex and RangeIndex.  I fixed this.  However, I think there should be a unit test in pysatNASA that ensures no data is loaded outside of the requested range.

**Comment by aburrell** [comment]
Also, the performance hit is big enough that I can no longer do an analysis I used to be able to do.  So, this effectively breaks pysatNASA for me.

**Comment by aburrell** [comment]
And by "this" it may be this PR or just "cdflib" in general.

**Comment by aburrell** [comment]
> Also, the performance hit is big enough that I can no longer do an analysis I used to be able to do. So, this effectively breaks pysatNASA for me.

Alright, things work again if I limit data loading to a day at a time instead of processing many days at once.  Given that, I am not sure where the biggest memory sink is.  But, importantly, this branch is working for me.  @JonathonMSmith once you add the try/except noted above (line 222, I believe), I think we can move forward with this PR.

**Comment by jklenzing** [comment]
Some error catching can be removed if MAVENSDC/cdflib#118 is merged.  We should update this in the code comments.

**Review by jklenzing — APPROVED** [review]
This looks good.  The larger cdflib style questions can be dealt with in a separate pull.  Created #53 to discuss whether we want to stash the cdflib routines elsewhere in the code.

**Pull request merged**